lib/semian.rb

```diff
@@ -175,10 +184,29 @@ def destroy(name)
   # the number of registered workers.
   # Semian.destroy removes the underlying resource, but
   # Semian.unregister will remove all references, while preserving
-  # the underlying semian resource (and sysV semaphore)
+  # the underlying semian resource (and sysV semaphore).
+  # Also clears any semian_resources
+  # in use by any semian adapters if the weak reference is still alive.
   def unregister(name)
     if resource = resources.delete(name)
-      resource.bulkhead.unregister_worker
+      resource.bulkhead.unregister_worker if resource.bulkhead
+      consumers_for_resource = consumers.delete(name) || []
+      consumers_for_resource.each do |consumer|
```

Comment:
Not arguing for anything necessarily, just trying to understand why it was written to support more than one consumer :)